File: lib/web_ui/skwasm/picture.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <emscripten.h>
#include "export.h"
#include "helpers.h"
#include "third_party/skia/include/core/SkPictureRecorder.h"
#include "wrappers.h"

using namespace Skwasm;

SKWASM_EXPORT SkPictureRecorder* pictureRecorder_create() {
  return new SkPictureRecorder();
}

SKWASM_EXPORT void pictureRecorder_dispose(SkPictureRecorder* recorder) {
  delete recorder;
}

SKWASM_EXPORT CanvasWrapper* pictureRecorder_beginRecording(
    SkPictureRecorder* recorder,
    const SkRect* cullRect) {
  return new CanvasWrapper{0, recorder->beginRecording(*cullRect)};
}

SKWASM_EXPORT SkPicture* pictureRecorder_endRecording(
    SkPictureRecorder* recorder) {
  return recorder->finishRecordingAsPicture().release();
}

SKWASM_EXPORT void picture_dispose(SkPicture* picture) {
  picture->unref();
}

SKWASM_EXPORT uint32_t picture_approximateBytesUsed(SkPicture* picture) {
  return static_cast<uint32_t>(picture->approximateBytesUsed());
}

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/picture.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
import 'package:ui/ui.dart' as ui;

class SkwasmPicture implements ui.Picture {
  SkwasmPicture.fromHandle(this._handle);
  final PictureHandle _handle;

  PictureHandle get handle => _handle;

  @override
  Future<ui.Image> toImage(int width, int height) {
    throw UnimplementedError();
  }

  @override
  void dispose() {
    ui.Picture.onDispose?.call(this);
    pictureDispose(_handle);
    debugDisposed = true;
  }

  @override
  int get approximateBytesUsed => pictureApproximateBytesUsed(_handle);

  @override
  bool debugDisposed = false;

  @override
  ui.Image toImageSync(int width, int height) {
    // TODO(jacksongardner): implement toImageSync
    throw UnimplementedError();
  }
}

class SkwasmPictureRecorder implements ui.PictureRecorder {
  factory SkwasmPictureRecorder() =>
    SkwasmPictureRecorder._fromHandle(pictureRecorderCreate());

  SkwasmPictureRecorder._fromHandle(this._handle);
  final PictureRecorderHandle _handle;

  PictureRecorderHandle get handle => _handle;

  void delete() => pictureRecorderDestroy(_handle);

  @override
  SkwasmPicture endRecording() {
    isRecording = false;

    final SkwasmPicture picture = SkwasmPicture.fromHandle(pictureRecorderEndRecording(_handle));
    ui.Picture.onCreate?.call(picture);
    return picture;
  }

  @override
  bool isRecording = true;
}
